### PR TITLE
more token in prefill

### PR DESCRIPTION
**Pull request by 3outeille**
much of FA3 improvements come from handling long sequences but the inputs were too small to take advantage of FA during prefill phase. Now making sure that the sequence is big enough during prefill phase and we see differnece of latency

```
=== Throughput & Memory Comparison (Single H100) ===
Model: HuggingFaceTB/SmolLM3-3B
Token budgets: [100]
Batch sizes: [16]
torch.utils.benchmark min_run_time=2.0s | mem_repeats=3
--------------------------------------------------------------------------------------------------------------------------------------------
Attention Implementation                  Batch   Tokens   Median Latency (s)     Tokens/s   Peak Alloc (GiB)  Peak Reserved (GiB)
--------------------------------------------------------------------------------------------------------------------------------------------
eager                    16      100           OOM/FAILED          N/A                N/A                  N/A
sdpa                     16      100                6.073        263.5              15.58                17.22
kernels-community/flash-attn     16      100                5.506        290.6              15.32                16.22
kernels-community/flash-attn3:flash_attention     16      100                5.496        291.1              15.32                16.22
--------------------------------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by ariG23498** [comment]
Took help of this PR to create a very lean reproduction using `torch.utils.benchmark.Timer.timeit`

```py
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"

from tqdm import tqdm
from torch import get_num_threads
from torch import inference_mode
from torch.utils.benchmark import Timer, Compare
from transformers import AutoModelForCausalLM, AutoTokenizer

MODEL_ID = "HuggingFaceTB/SmolLM3-3B"
DEVICE = "cuda:0"
BATCH_SIZE = 16
MAX_INPUT_LENGTH = 4096

tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
print(f"PAD Token for tokenizer = {tokenizer.pad_token}")

base_prompts = [
    "What is Tensor Parallelism?",
    "Explain machine learning fundamentals.",
    "How do neural networks work?",
    "What are the benefits of distributed computing?",
    "Describe the attention mechanism in transformers.",
    "What is gradient descent?",
    "How does backpropagation work?",
    "Explain the concept of overfitting.",
]
messages = [
    [{"role": "system", "content": base_prompts[i % len(base_prompts)]}] for i in range(BATCH_SIZE)
]
texts = [tokenizer.apply_chat_template(m, add_generation_prompt=True, tokenize=False) for m in messages]

inputs = tokenizer(
    texts,
    return_tensors="pt",
    padding="max_length",
    truncation=True,
    max_length=MAX_INPUT_LENGTH,
).to(DEVICE)


@inference_mode
def run_inference(model, tokenizer, inputs, max_new_tokens):
    outputs = model.generate(
        **inputs,
        do_sample=False,
        temperature=None,
        max_new_tokens=max_new_tokens,
        eos_token_id=-1,
        disable_compile=True,
        use_cache=True,
        return_dict_in_generate=True,  # so we can count actual generated length
    )
    return outputs

# Configurations: (description, attn_implementation)
configs = [
    ("eager attention", "eager"),
    ("sdpa attention", "sdpa"),
    ("flash attention", "kernels-community/flash-attn"),
    ("flash 3 attention", "kernels-community/flash-attn3:flash_attention"),
]

results = []
label = "Time taken to generate tokens"
max_new_tokens_list = [32, 64, 128]

for description, attn_impl in configs:
    model = AutoModelForCausalLM.from_pretrained(
        MODEL_ID,
        dtype="auto",
        attn_implementation=attn_impl,
    ).to(DEVICE)
    
    for max_new_tokens in tqdm(max_new_tokens_list):
        results.append(
            Timer(
                stmt="run_inference(model, tokenizer, inputs, max_new_tokens)",
                setup='from __main__ import run_inference',
                globals={"model": model, "tokenizer": tokenizer, "inputs": inputs, "max_new_tokens": max_new_tokens},
                num_threads=get_num_threads(),
                label=label,
                sub_label=f"num tokens: {max_new_tokens}",
                description=description
            ).timeit(5)
        )
    
    model.to("cpu")
    del model

compare = Compare(results)
compare.print()
```

The results are aligned to the findings of this PR:

```
[----------------------- Time taken to generate tokens -----------------------]
                       |  eager attention  |  sdpa attention  |  flash attention  |  flash 3 attention
11 threads: ------------------------------------------------------------------------------------------
      num tokens: 32   |        4.7        |       2.9        |        2.3        |         2.3       
      num tokens: 64   |        6.2        |       4.4        |        3.7        |         3.7       
      num tokens: 128  |        9.2        |       7.4        |        6.6        |         6.6       

Times are in seconds (s).
```
